spring-web/src/main/java/org/springframework/http/converter/AbstractHttpMessageConverter.java

```diff
@@ -224,11 +227,23 @@ public HttpHeaders getHeaders() {
 			}));
 		}
 		else {
-			writeInternal(t, outputMessage);
+			writeInternal(unboxed, outputMessage);
 			outputMessage.getBody().flush();
 		}
 	}
 
+	/**
+	 * Unbox the object value if it is stored in a container (like {@link java.util.Optional}.
+	 */
+	@SuppressWarnings("unchecked")
+	protected T unboxIfNeeded(T value) {
```

Comment:
Quick question about the unwrapping and returning type parameter: basically what ur doing here is convert a wrapped (Optional) value of type `T` to an unwrapped value and in the process of this changing the returned type. 
How are you going to be ensuring type safety? Especially for the abstract method `writeInternal` which may be implemented by anyone?

Let's say that you have type argument `Optional<String>` and the `unboxIfNeeded` method converts this type to `String`. This all while `writeInternal` and the rest of the code still expect `T` to be `Optional<String>`, this especially is an issue since you tell the compiler to expect type safety (by suppressing the warning). 
These seem like mismatching types which might, depending on the implementing code, throw a ClassCastException. This may fix the issue but you will now not know what the type of variable `t` really is.

_(I am fully aware that using Optional as a type argument is a bad practice, however I would value type safety over implementation of the user)_

Reply:
Correct, in case when a converter has generic type `Optional<*>`, it would fail with the class cast exception. I will fix that

Reply:
I see that you are now "ensuring" type safety. However, there are still some issues which resolve around the second part of my comment.

In my comment I mentioned the following:
> This may fix the issue, but you will now not know what the type of variable t really is going to be. When you implement AbstractHttpMessageConverter by yourself, you are going to be having restrictions for a fix on Jackson's implementation. I would suggest not changing this class but look at the Jackson specific implementation (AbstractJackson2HttpMessageConverter) since you won't be needing the type safety here and it tackles this specific issue.

With this, I was referring to the strategy pattern used by these classes (`HttpMessageConverter`, `GenericHttpMessageConverter`, `AbstractHttpMessageConverter` and `AbstractGenericHttpMessageConverter`). These classes are to be implemented and allow the user to define, at run-time, the to be chosen strategy. 
Since this is the case, we do not want the user to be confused or to already have some sort of pre-processing on any implementation that has been made. Because this specific issue (the unboxing issue) lies within the `AbstractJackson2HttpMessageConverter` class and sub classes: why not fix it there as it will fit that specific use case? Now we are going to be introducing bugs and unexpected behavior for anyone implementing this base class in it's own converter, which is most certainly not what we would want.

Reply:
As agreed here #24867, we want to avoid code duplication. You are still able to override the `unboxIfNeeded` in the impl in case you really don't want the `Optional` to be unwrapped